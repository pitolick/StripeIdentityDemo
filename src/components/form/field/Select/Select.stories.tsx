import type { Meta, StoryObj } from '@storybook/react'
import { Select } from './Select'

const meta: Meta<typeof Select> = {
  title: 'Form/Field/Select/Base',
  component: Select,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof Select>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
  },
}

/**
 * プレースホルダー
 */
export const Placeholder: Story = {
  args: {
    options: [
      { value: '', label: '選択してください', disabled: true },
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    defaultValue: '',
  },
}

/**
 * 初期値
 */
export const InitialValue: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    defaultValue: '2',
  },
}

/**
 * マルチセレクト
 */
export const MultiSelect: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    multiple: true,
  },
}
