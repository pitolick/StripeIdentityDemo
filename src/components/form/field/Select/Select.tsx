/**
 * セレクトボックス非制御コンポーネント
 * @module Select
 */
import React, { forwardRef } from 'react'

/**
 * セレクトボックスコンポーネントのプロパティ
 * @typedef {Object} SelectProps
 * @property {string} options - セレクトボックスに表示するテキスト
 * @property {boolean} full - ブロック要素化
 * @property {React.SelectHTMLAttributes<HTMLSelectElement>} - セレクトボックスタグのHTML属性
 */
export interface SelectProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  /**
   * 子要素
   */
  options: React.OptionHTMLAttributes<HTMLOptionElement>[]
  /**
   * 初期値
   */
  defaultValue?: string
}

/**
 * セレクトボックス非制御コンポーネント
 * @param {SelectProps} props - セレクトボックスコンポーネントのプロパティ
 * @returns {JSX.Element} - セレクトボックスコンポーネント
 */
export const Select = forwardRef<HTMLSelectElement, SelectProps>((props, ref) => {
  const { options, className = undefined, defaultValue, ...selectProps } = props
  const classNameProps = className ? className : ''
  return (
    <>
      <select
        ref={ref}
        defaultValue={defaultValue}
        {...selectProps}
        className={`${classNameProps} block w-full rounded-md border-gray-500 py-2 pl-3 pr-7 text-sm md:py-3 md:pl-5 md:pr-9 md:text-base`}
      >
        {options.map((option, index) => {
          const { label, ...optionAttr } = option
          return (
            <option key={`${option.value}_${index}`} label={label} {...optionAttr}>
              {label}
            </option>
          )
        })}
      </select>
    </>
  )
})

Select.displayName = 'Select'
