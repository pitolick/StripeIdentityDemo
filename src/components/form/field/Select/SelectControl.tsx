/**
 * セレクトボックス制御コンポーネント（react-hook-form）
 * @module SelectControl
 */
import React, { forwardRef } from 'react'
import { Select, SelectProps } from './Select'
import { InputBox, InputBoxProps } from '@/components/form/parts'

/**
 * セレクトボックスコンポーネントのプロパティ
 * @memo バリデーションスキーマ（zod）は各入力子コンポーネント内でexportしformコンポーネント内でresolverに指定する
 * @typedef {Object} SelectControlProps
 */
export interface SelectControlProps extends SelectProps, Omit<InputBoxProps, 'children'> {}

/**
 * セレクトボックス制御コンポーネント（react-hook-form）
 * @param {SelectControlProps} props - セレクトボックスコンポーネントのプロパティ
 * @returns {JSX.Element} - セレクトボックスコンポーネント
 */
export const SelectControl = forwardRef<HTMLSelectElement, SelectControlProps>((props, ref) => {
  const { nameId, label, defaultValue, note, required = undefined, options, className, error, contentWidth, ...selectProps } = props
  const classNameProps = className ? className : ''
  // エラー時のclass名
  const errorClass = error ? `bg-red-pale outline outline-red outline-2 -outline-offset-2` : ''
  return (
    <InputBox nameId={nameId} label={label} note={note} required={required} error={error} contentWidth={contentWidth}>
      <Select ref={ref} id={nameId} options={options} defaultValue={defaultValue} className={`${classNameProps} ${errorClass}`} {...selectProps} />
    </InputBox>
  )
})

SelectControl.displayName = 'SelectControl'
