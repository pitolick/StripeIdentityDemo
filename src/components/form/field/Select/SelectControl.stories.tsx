import type { Meta, StoryObj } from '@storybook/react'
import { SelectControl } from './SelectControl'

const meta: Meta<typeof SelectControl> = {
  title: 'Form/Field/Select/Control',
  component: SelectControl,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof SelectControl>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'base',
  },
}

/**
 * プレースホルダー
 */
export const Placeholder: Story = {
  args: {
    options: [
      { value: '', label: '選択してください', disabled: true },
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    defaultValue: '',
    nameId: 'placeholder',
  },
}

/**
 * 初期値
 */
export const InitialValue: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    defaultValue: '2',
    nameId: 'initial-value',
  },
}

/**
 * ラベル
 */
export const Label: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'label',
    label: 'ラベル',
  },
}

/**
 * 注釈
 */
export const Note: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'note',
    label: 'ラベル',
    note: { list: ['注釈'] },
  },
}

/**
 * バリデーション：必須
 */
export const Required: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'required',
    label: 'ラベル',
    required: true,
  },
}

/**
 * バリデーション：任意
 */
export const Optional: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'optional',
    label: 'ラベル',
    required: false,
  },
}

/**
 * バリデーション：エラー
 */
export const Error: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'error',
    label: 'ラベル',
    note: { list: ['注釈'] },
    required: true,
    error: { type: 'required', message: '必須項目です' },
  },
}

/**
 * マルチセレクト
 */
export const MultiSelect: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'multi-select',
    label: 'ラベル',
    note: { list: ['注釈'] },
    required: true,
    multiple: true,
  },
}

/**
 * マルチセレクト：エラー
 */
export const MultiSelectError: Story = {
  args: {
    options: [
      { value: '1', label: '選択肢1' },
      { value: '2', label: '選択肢2' },
      { value: '3', label: '選択肢3' },
    ],
    nameId: 'multi-select-error',
    label: 'ラベル',
    note: { list: ['注釈'] },
    required: true,
    multiple: true,
    error: { type: 'required', message: '必須項目です' },
  },
}
