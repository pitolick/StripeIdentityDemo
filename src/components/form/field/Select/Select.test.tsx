import React from 'react'
import { render } from '@testing-library/react'
import { Select, SelectProps } from './Select'

describe('Select コンポーネント', () => {
  const options: React.OptionHTMLAttributes<HTMLOptionElement>[] = [
    { value: 'option1', label: 'Option 1' },
    { value: 'option2', label: 'Option 2' },
    { value: 'option3', label: 'Option 3' },
  ]

  const defaultProps: SelectProps = {
    options,
    defaultValue: 'option1',
  }

  it('正しくレンダリングされること', () => {
    const { container } = render(<Select {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  it('classNameプロパティが指定された場合、正しく適用されること', () => {
    const { container } = render(<Select {...defaultProps} className="text-red" />)
    expect(container).toMatchSnapshot()
  })
})
