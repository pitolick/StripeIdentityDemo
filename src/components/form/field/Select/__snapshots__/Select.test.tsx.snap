// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select コンポーネント classNameプロパティが指定された場合、正しく適用されること 1`] = `
<div>
  <select
    class="text-red block w-full rounded-md border-gray-500 py-2 pl-3 pr-7 text-sm md:py-3 md:pl-5 md:pr-9 md:text-base"
  >
    <option
      label="Option 1"
      selected=""
      value="option1"
    >
      Option 1
    </option>
    <option
      label="Option 2"
      value="option2"
    >
      Option 2
    </option>
    <option
      label="Option 3"
      value="option3"
    >
      Option 3
    </option>
  </select>
</div>
`;

exports[`Select コンポーネント 正しくレンダリングされること 1`] = `
<div>
  <select
    class=" block w-full rounded-md border-gray-500 py-2 pl-3 pr-7 text-sm md:py-3 md:pl-5 md:pr-9 md:text-base"
  >
    <option
      label="Option 1"
      selected=""
      value="option1"
    >
      Option 1
    </option>
    <option
      label="Option 2"
      value="option2"
    >
      Option 2
    </option>
    <option
      label="Option 3"
      value="option3"
    >
      Option 3
    </option>
  </select>
</div>
`;
